### ISSUE TITLE
command line substitution for optional inputs

### ISSUE THREAD

**Issue by leej3** [boutiques/boutiques#303]
@gkiar can't remember if this is neatly facilitated. Current hacky solution is the specify both has optional and hope one is provided.

**Comment by glatard** [comment]
sounds like a serious issue but a brief example would help understanding exactly what it is about. Are you trying to have a group of inputs among which at least one is required?

**Comment by gkiar** [comment]
Hey @glatard - the situation is as follows:

- We have a required input, `param1`
- We have an optional input, `opt1`
- We have an output file that will have the filename `param1-thing.txt` if `opt1` is not specified, but if `opt1` is specified, it will be `opt1-thing.txt`, where in both cases `param1` and `opt1` are replaced with the values assigned to them.

My current understanding is I would need to specify two optional output files, one for each of these cases, and then hopefully one of my two outputs would exist, but I wouldn't be able to validate that since they're both optional.

Not unlike this, I'd also like to map the input file extension to the output file --- after stripping extensions, can we also replace the stripped extension (when it was one on a list of many options) back onto new outputs?

**Comment by glatard** [comment]
ok, got it! Your current understanding is correct, unfortunately :) 
So we need to specify when optional outputs will be produced: do we just want a boolean expression, or are we up for content-based dependencies? 
It also reminds me of a discussion we had with @satra in August, they solved it in Nipype using a matrix if I remember well, perhaps talk with him if he's around?

**Comment by satra** [comment]
we will be having a discussion about this after lunch today.

**Comment by leej3** [comment]
@gkiar  and I had another talk about this. One problem that we would need to solve to fully specify the output filepaths is the ability to reuse stripped extensions. So for example with:

```
"path-template-stripped-extensions": [
              ".HEAD", ".BRICK", ".",
              "+orig", "+acpc", "+tlrc"
            ]
```

We could specify certain options that add elements of this list. Something along the lines of:

```
"output-files": {
  ... , 
  "optional": false,
  "path-template":
  [
    {"opt2 > 10":            "[PARAM1]-thing_large.txt"},
    {"default":              "[PARAM1]-thing.txt"},
    {"opt3":                  "[PARAM1]-thing[STRIPPED_EXTENSIONS[3:]].txt
  ]
}
```
That would add   "+orig", "+acpc", or "+tlrc" if they were found in the relevant input value.

**Comment by satra** [comment]
and just to keep other pieces in mind. also to have non-file outputs like dictionaries and simple literals.

**Comment by glatard** [comment]
for the output naming issue, I'm worried that using our own Python-inspired grammar might create issues down the road: usability for non Python developers, limitations in expressiveness (do we allow *any* Python expression?), issues with deployment (parsing the descriptor needs Python). 

Instead, what do you think of providing an executable in the container to encode this logic? Syntax would have to be `<some_name_provided_in_the_descriptor> <invocation> --<output_id>` and it would print the output file name(s) on the stdout. Motivation: sh is the common interoperability ground, not Python. It's a bit overheady but at the same time, Boutiques is about CLI tools, so why not push it to the limit?

**Comment by gkiar** [comment]
From discussion with @glatard:

 - We will support logic written with Python syntax
 - Expressions will be organized like shown in https://github.com/boutiques/boutiques/issues/303#issuecomment-421434734, with  and will be evaluated in an "if, elseif" order, where the first hit is the only one activated.
 - The stripped extensions would apply to all conditions uniformly, where the stripping is applied sequentially to all fields (i.e. can remove `.HEAD` and then `+orig` from an input)

**Comment by glatard** [comment]
From discussion at the meeting today:
* Do not backward compatibility: separate `conditional-path-templates` to the regular `path-template`. Make fields mutually exclusive. 
* Default is required if output is required
* Ordering should be if-elseif, go linearly
* All extensions are stripped in all path templates

Wait for a few weeks or months before releasing when ready, to allow for feedback.

**Comment by jerdra** [comment]
hey @glatard this is similar to an issue that we're having (templating based on other arguments) but with an additional requirement that we'd like to be able to substitute values into an invocation. 

For example when running a command-line tool that takes a particular participant name (let's say BIDS) as an argument it doesn't make sense to encode that into the invocation.json as a static value - as we'd have to have a single invocation.json for each BIDS participant. I'm aware that a list of subjects may be provided in some cases - but this isn't practical when submitting jobs to a cluster; where we usually submit 1 participant/job. Our current solution is to produce intermediate invocations where a subject field is injected - but this is hacky. 

In addition, other arguments may rely on knowing the participant ID (such as providing additional optional files [SUBJECT].surf.gii etc..) - which is what I presume is being addressed in this issue. 

Would it make sense to open up some arguments to be either outside the invocation json or as a variable in the invocation json (e.g using a list of subjects that will substitute the subject variable in the invocation)? 

Let me know if it would make sense to produce a separate issue for this - i figured it was related since this discussion is somewhat related to variable inputs.

**Comment by glatard** [comment]
This is now available in version 0.5.23, thanks @DarrinFong!
